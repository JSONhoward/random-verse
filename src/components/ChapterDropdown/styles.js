import styled from 'styled-components'

export const DropdownDiv = styled.div`
    position: relative;
    display: flex;
    flex-direction: row;
    align-items: center;
    /* justify-content: center; */
    height: 4rem;
    width: 10rem;
    padding-left: 1rem;
    margin: 2px;
    box-shadow: 2px 2px 5px black;
    background-color: white;
    font-size: 1.5rem;

    @media (max-width: 500px) {
        height: 3rem;
        width: 5rem;
        font-size: .75rem;
        padding-left: .5rem;
    }
    `

export const OptionsDiv = styled.div`
    position: absolute;
    top: 65px;
    left: 0px;
    display: ${props => props.isOpen ? 'flex' : 'none'};
    flex-direction: column;
    align-items: left;
    justify-content: top;
    height: fit-content;
    max-height: 10rem;
    width: 7.8rem;
    background-color: white;
    animation: fadeIn linear .5s;
    box-shadow: 2px 2px 5px black;
    padding: 2px;
    overflow-y: scroll;
    z-index: 5;
    cursor: pointer;

    @media (max-width: 500px) {
        width: 5rem;
        font-size: 1rem;
        top: 50px;
    }
    `

export const ListItem = styled.div`
    margin-bottom: 2px;
    text-align: center;
    font-style: italic;
    `

export const Caret = styled.div`
    position: absolute;
    right: 0;
    display: flex;
    align-items: center;
    justify-content: center;
    height: 100%;
    width: 2rem;
    background: linear-gradient(to bottom, whitesmoke, lightgrey);
    cursor: pointer;

    @media (max-width: 500px) {
        width: 1rem;
    }
    `