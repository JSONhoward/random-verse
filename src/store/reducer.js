export const reducer = (state, action) => {
    let book, chapter, verse, toggle, displayVerse
    switch(action.type) {
        case 'LOADING_TRUE':
            state = {...state, loading: true}
            return state;
        case 'LOADING_FALSE':
            state = {...state, loading: false}
            return state;
        case 'SET_VERSE':
            displayVerse = action.payload
            state = {...state, displayVerse}
            return state;
        case 'TOGGLE_BOOK':
            toggle = !state.bookOptions
            state = {...state, bookOptions: toggle}
            return state;
        case 'TOGGLE_CHAPTER':
            toggle = !state.chapterOptions
            state = {...state, chapterOptions: toggle}
            return state;
        case 'TOGGLE_VERSE':
            toggle = !state.verseOptions
            state = {...state, verseOptions: toggle}
            return state;
        case 'CHOOSE_BOOK':
            book = action.payload
            chapter = '(Random)'
            verse = '(Random)'
            state = {...state, book, chapter, verse, bookOptions: false}
            return state;
        case 'CHOOSE_CHAPTER':
            chapter = action.payload
            state = {...state, chapter, chapterOptions: false}
            return state;
        case 'CHOOSE_VERSE':
            verse = action.payload
            state = {...state, verse, verseOptions: false}
            return state;
        case 'CLOSE_ALL':
                state = {
                    ...state,
                    bookOptions: state.bookOptions && false,
                    chapterOptions: state.chapterOptions && false,
                    verseOptions: state.verseOptions && false
                }
            return state;
        default:
            return state;
    }
}