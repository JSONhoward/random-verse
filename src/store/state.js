import {createContext} from 'react'

const homeContent = 'For God so loved the world, that he gave his only begotten Son, that whosoever believeth in him should not perish, but have everlasting life.'
const homeVerse = 'John 3:16'

export const initialState = {
    bookOptions: false,
    chapterOptions: false,
    verseOptions: false,
    book: '(Random)',
    chapter: '(Random)',
    verse: '(Random)',
    displayVerse: {
        verse: homeVerse,
        content: homeContent
    },
    loading: false
}

export const StateContext = createContext(null)