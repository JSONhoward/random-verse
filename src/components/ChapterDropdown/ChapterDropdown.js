import React, { useContext } from 'react'
import { FaCaretDown } from 'react-icons/fa'

import { StateContext } from '../../store/state'
import { BIBLE } from '../../utils/bible'
import {ListItem, DropdownDiv, OptionsDiv, Caret} from './styles'

const ChapterDropdown = () => {
    const [state, dispatch] = useContext(StateContext)
    const { chapterOptions, book, chapter } = state

    const chapterList = book !== "(Random)" ? BIBLE.filter(el => {
        return el.name === book
    }).map(el => el.chapters).flat(1) : []

    let dropdownList = chapterList.map((item, i) => {
        return (
            <ListItem
                tabIndex={1}
                id={i}
                onClick={() => chooseChapter(i)}
                key={i}
                onKeyPress={e => {
                    if (e.charCode === 32 || e.charCode === 13) {
                        chooseChapter(i)
                    }
                }}>
                Ch. {i + 1}
            </ListItem>
        )
    })

    const chooseChapter = option => {
        dispatch({ type: 'CHOOSE_CHAPTER', payload: option + 1 })
    }

    const keyOpen = (e) => {
        if (e.target.id === 'dropdown1') {
            if (e.charCode === 13 || e.charCode === 32) {
                dispatch({ type: 'TOGGLE_CHAPTER' })
            }
        }
    }

    const clickOpen = () => {
        dispatch({ type: 'TOGGLE_CHAPTER' })
    }

    return (
        <>
            <DropdownDiv id={'dropdown1'} onKeyPress={e => keyOpen(e)} tabIndex={1}>
                <p>{chapter}</p>
                <Caret onClick={clickOpen}>
                    <FaCaretDown size={'1.5rem'} />
                </Caret>
                <OptionsDiv isOpen={chapterOptions}>
                    {dropdownList}
                </OptionsDiv>
            </DropdownDiv>
        </>
    )
}

export default ChapterDropdown
