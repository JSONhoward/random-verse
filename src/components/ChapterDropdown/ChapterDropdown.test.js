import React from 'react'
import {create} from 'react-test-renderer'
import { initialState, StateContext } from '../../store/state'
import {reducer} from '../../store/reducer'
import ChapterDropdown from './ChapterDropdown'

describe('ChapterDropdown Tests', () => {
    test('Snapshot Test', () => {
        const Component = () => {
            const [state, dispatch] = React.useReducer(reducer, initialState)
            return(
                <StateContext.Provider value={[state, dispatch]}>
                    <ChapterDropdown />
                </StateContext.Provider>
            )
        }

        const tree = create(<Component />).toJSON()

        expect(tree).toMatchSnapshot()
    })
})