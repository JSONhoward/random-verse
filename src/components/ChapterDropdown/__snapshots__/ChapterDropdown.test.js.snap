// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChapterDropdown Tests Snapshot Test 1`] = `
<div
  className="sc-AxjAm glzBBN"
  id="dropdown1"
  onKeyPress={[Function]}
  tabIndex={1}
>
  <p>
    (Random)
  </p>
  <div
    className="sc-AxhCb OGtmh"
    onClick={[Function]}
  >
    <svg
      fill="currentColor"
      height="1.5rem"
      size="1.5rem"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 320 512"
      width="1.5rem"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
      />
    </svg>
  </div>
  <div
    className="sc-AxirZ cOlIR"
  />
</div>
`;
